src/PureScript/Backend/Optimizer/Syntax.purs

```diff
@@ -51,7 +51,7 @@ sndPair (Pair _ a) = a
 data BackendAccessor
   = GetProp String
   | GetIndex Int
-  | GetOffset Int
+  | GetCtorField (Qualified ProperName) (Qualified Ident) Int
```

Comment:
Would it be possible to mirror the fields in CtorSaturated? Possibly replacing the array of fields with the field name as well as offset?

Reply:
Assuming you want this representation:
```purs
--           type name              ctor type       ctor name         offset & field name
GetCtorField (Qualified ProperName) ConstructorType (Qualified Ident) (Tuple Int String)
```

It appears that I can get the constructor's field names from the `ConvertEnv`'s `datatypes` map. However, I think that only includes constructors defined in the current module, not those imported from elsewhere. Do I use `implementations` or `moduleImplementations` to get that additional info?

If instead, you want this representation, which maps more 1-to-1 with `CtorSaturated`...
```purs
--           module name?       ctor type       ty name   ctor name   offset & field name
GetCtorField (Qualified Ident) ConstructorType ProperName Ident      (Tuple Int String)
```
we'll need to pull more info out of `ConvertEnv`.

Reply:
Given
```purescript
--            qualified ctor    ctor type       type name  tag   fields
CtorSaturated (Qualified Ident) ConstructorType ProperName Ident (Array (Tuple String a))
```
I would want something like
```purescript
GetCtorField (Qualified Ident) ConstructorType ProperName Ident String Int
```
I don't think it's necessary to embed a Tuple.

> Do I use implementations or moduleImplementations to get that additional info?

I think that you would use both. `implementations` is the transitive implementations, `moduleImplementations` are the current module's implementations.

Reply:
`moduleImplementations` appears to be iteratively-updated as the bindings are converted to `BackendExpr`, which means a type that is declared after it is used wouldn't be in `moduleImplementations` at the time I need it, right?
```purs
useFoo :: Foo -> Int
useFoo (Foo i) = i

data Foo = Foo Int
```

`dataTypes` on the other hand is calculated before the bindings are converted, so I opted to use that for the current modules data types.